controllers/user.controller.js

```diff
@@ -0,0 +1,157 @@
+      const NewUser = new User(user);
+      const salt = await bcrypt.genSalt(10);
+      NewUser.password = await bcrypt.hash(NewUser.password, salt);
+      await NewUser.save();
+
+      NewUser.__v = undefined;
+      NewUser.password = undefined;
+
+      res.status(200).json({ message: "user added", NewUser });
+   } catch (error) {
+      console.log(error);
+      if (error.name === "MongoError" && error.code === 11000) {
+         if (error.keyPattern.userName) {
+            return res.status(422).json({
+               message: "user name is already taken",
+               errorMessage: error.message,
+            });
+         } else if (error.keyPattern.email) {
+            return res.status(422).json({
+               message: "account already exists with this email",
+               errorMessage: error.message,
+            });
+         }
+      }
+      res.status(500).json({
+         message: "could not add user",
+         errorMessage: error.message,
+      });
+   }
+};
+
+const loginUser = async (req, res) => {
+   try {
+      const { username, password } = req.body;
+
+      let user = await User.findOne({ userName: username });
+
+      if (user) {
+         const verifyPassword = await bcrypt.compare(password, user.password);
+         if (verifyPassword) {
+            const token = jwt.sign({ userId: user._id }, JWT_KEY, { expiresIn: "24h" });
+            user.__v = undefined;
+            user.password = undefined;
+            return res.status(200).json({ message: "user logged in", user, token });
+         }
+         return res.status(403).json({ message: "username and password did not match" });
+      }
+      res.status(404).json({ message: "user does not exist " });
+   } catch (error) {
+      res.status(500).json({
+         message: "cannot retrieve user",
+         errorMessage: error.message,
+      });
+   }
+};
+
+const resetOrUpdateUserPassword = async (req, res) => {
+   try {
+      const { email, password } = req.body;
+      const updateUserPassword = { password: password };
+      let user = await User.findOne({ email: email });
+
+      if (!user) {
+         return res.status(404).json({ message: "user does not exist" });
+      }
+      user = extend(user, updateUserPassword);
+      const salt = await bcrypt.genSalt(10);
+      user.password = await bcrypt.hash(user.password, salt);
+      user = await user.save();
+      res.status(200).json({ message: "user credentials updated", user });
+   } catch (error) {
+      res.status(500).json({
+         message: "cannot retrieve user",
+         errorMessage: error.message,
+      });
+   }
+};
+
+const getUserProfile = (req, res) => {
+   const { user } = req;
+   user.__v = undefined;
+   res.status(200).json({ user });
+};
+
+const updateUserProfile = async (req, res) => {
+   try {
+      let { user } = req;
+      const { password } = req.body;
+      const updateUserPassword = { password: password };
```

Comment:
With ES6, you need to not repeat `Key: value` if both are the same :)! Just a suggestion

Reply:
yes, you are right. done!